### ISSUE TITLE
Injecting a config file at startup through DI

### ISSUE THREAD

**Issue by nismolo** [LightBDD/LightBDD#85]
Hey,

We are trying to implement a mechanism to have a configuration file which consists of properties of specific environments for example, a staging environment config json file we will have the environment url, test user credentials etc.. 

We are trying to achieve this by injecting the configuration file within the OnConfigure(LightBddConfiguration configuration) method which will then through dependency injection and inversion of control make the file accessible throughout the application.  

But we are having difficulties accessing the configuration service from the  constructor of the feature steps class since the feature steps class and the feature class are have different constructor parameters.

I went through the LightBDD Configuration documentation but I cannot find any information related to how to implement startup property configuration which would be great for an E2E testing framework. Do you have an example in which IOC is being used?

Also on the documentation it seems that there is no reference to the onSetUp and onTearDown methods of the LightBddScopeAttribute class. Maybe i didn't manage to find them :) but when I tested them out the code within the methods actually never got executed.

**Comment by Suremaker** [comment]
Hello,

The quick answer to your question on DI support in LightBDD is that LightBDD does not provide any special support to DI so you have to relay on what underlying testing framework offers you.

Xunit offers a very limited support here (see https://xunit.github.io/docs/shared-context.html):
* ICollectionFixture that will make a dependency shared between all the test classes (it also means that those classes cannot be tested in parallel :( - which usually is a big drawback for E2E tests ). I tried to make a working sample for you with it, but somehow failed. If you are interested in that more, we could spend some time investigating it further, but as I mentioned it means no parallel execution for classes sharing collection,
* IClassFixture that will share the same instance of the dependency between all tests belonging to given class.

The other syntax works (I was successfully using it for tests with Selenium drivers for example), and it looks as follows:

```c#
public partial class My_feature : FeatureFixture, IClassFixture<Dependency>
{
    public My_feature(ITestOutputHelper output, Dependency dependency)
        : base(output)
    {
        // ...
    }
}
```

You can always play with it more, however I am not sure if that would be useful for you.

-------------

Having said that, I was usually using LightBDD for running acceptance tests against live service where on setup I was starting it, running all the tests and killing it at the very end.
This is the similar scope to you DI container as what I understand you would like to initialize it once and use everywhere.

Normally xunit does not offer you any hook for initializing code before all tests and terminating it after all are done, but this is the reason why I added OnSetUp() / OnTearDown() methods to `LightBddScopeAttribute`, where usage is as follows:

```c#
// here we are using our customized scope - perhaps you missed that by saying that this code was never executed ;-)
[assembly: ConfiguredLightBddScope]

namespace LightBddWithIoc
{
    public class ConfiguredLightBddScope : LightBddScopeAttribute
    {
        protected override void OnSetUp()
        {
            DependencyContainer.Initialize();
        }

        protected override void OnTearDown()
        {
            DependencyContainer.Terminate();
        }
    }
}
```

So, the above code gives us a way to configure container and dispose it at the end of execution.

The most simplistic implementation of the container then could look like that:
```c#
public class DependencyContainer
{
    public static IContainer Container { get; private set; }

    public static void Initialize()
    {
        var builder = new ContainerBuilder();
        builder.RegisterType<MyLogic>().As<IMyLogic>().SingleInstance();
        // more registration ...
        Container = builder.Build();
    }

    public static void Terminate()
    {
        Container?.Dispose();
    }
}
```

... and the simplest and quickest usage of it in feature can look as follows:
```c#
public partial class My_feature : FeatureFixture
{
    private IMyLogic _myLogic;

    public My_feature(ITestOutputHelper output)
        : base(output)
    {
    }

    private void Given_IMyLogic_instance()
    {
        // the IoC container is accessed here statically
        _myLogic = DependencyContainer.Container.Resolve<IMyLogic>();
    }

    private void Then_calling_Hello_method_on_it_should_return_value(string value)
    {
        Assert.Equal(value, _myLogic.Hello());
    }
}
```

The solution presented above is the simplistic one where IoC container is accessed statically in the code.

Unfortunately I am not aware of any way to make xunit accepting `public My_feature(ITestOutputHelper output, IMyLogic myLogic)` constructor and resolving the dependency with IoC usage - which I believe you are looking for.

I think you can however wrap the static access to the container by:
* making a base fixture class with `IContainter Container => DependencyContainer.Container;` property,
* wrapping it in class and pass as `IClassFixture<Wrapper>` to the fixtures

... but all of those approaches would be more like syntactic sugar hiding the fact that we are dealing with static container (maybe with the exception of first option that could make code using the container shorter).

---------------
Finally I have attached a sample project showing it: 
[lightbddioc.zip](https://github.com/LightBDD/LightBDD/files/1489309/lightbddioc.zip) where:
* My_feature class shows usage of static container reference
* My_feature_with_container_resolver_wrapper presents more the usage of `IClassFixture<>`, as I will rather not use it to wrap the container.

I hope it would help and feel free to let me know if you have any questions or issues with it.

Ps. It looks like I did not add any documentation regarding LightBddScope OnSetUp / OnTearDown methods. Thanks for pointing it out - I'll add it soon.

**Comment by nismolo** [comment]
Thanks @Suremaker for your prompt reply. We are currently using a temporary solution by extending the FeatureFixture and setting the configuration data in it and making it available to all feature steps class., We will look into the issue in more detail and try to implement one of your suggested solutions.

We are currently looking into another issue which is making the tests run dynamically according to the passed arguments when initializing the test suite from command line but it seems xunit.runner.console does not allow custom commands... :( https://github.com/xunit/xunit/issues/381

**Comment by nismolo** [comment]
Hey we implemented the IClassFixture solution you suggested and we managed to get everything up and running according to plan. Thanks mate. 

As for the issue of passing arguments we are currently using environment variables and setting them up through Bash its not ideal but its doing the job for now. .

**Comment by Suremaker** [comment]
Cool, it is nice to hear that.

I would love to hear your feedback on LightBDD features - i.e. what do you like, what is not that good and if there is anything that should be changed or added :)
